libWlz/WlzGreyValue.3
'\" te
.\" ident MRC HGU $Id$
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" Project:    Woolz
.\" Title:      WlzGreyValue.3
.\" Date:       July 1997
.\" Author:     Bill Hill
.\" Copyright:	1997 Medical Research Council, UK.
.\"		All rights reserved.
.\" Address:	MRC Human Genetics Unit,
.\"		Western General Hospital,
.\"		Edinburgh, EH4 2XU, UK.
.\" Purpose:    Woolz functions for random access to grey values.
.\" $Revision$
.\" Maintenance:Log changes below, with most recent at top of list.
.\"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.nh
.TH "WlzGreyValue" 3 "MRC HGU Woolz" "Woolz Procedure Library"
.SH NAME
WlzGreyValueMakeWSp,
WlzGreyValueFreeWSp,
WlzGreyValueGet,
WlzGreyValueGetCon \- Woolz functions for random access to grey values.
.SH SYNOPSIS
.LP
.B #include <Wlz.h>
.LP
.BI "WlzGreyValueWSpace *WlzGreyValueMakeWSp(WlzObject *" "obj" ,
.in +4m
.br
.BI "WlzErrorNum *" "dstErrNum" );
.in -4m
.LP
.BI "void WlzGreyValueFreeWSp(WlzGreyValueWSpace *" "gVWSp" );
.LP
.BI "void WlzGreyValueGet(WlzGreyValueWSpace *" "gVWSp" ,
.in +4m
.br 
.BI "double " "plane" ,
.BI "double " "line" ,
.BI "double " "kol" );
.in -4m
.LP 
.BI "void WlzGreyValueGetCon(WlzGreyValueWSpace *" "gVWSp" ,
.in +4m
.br 
.BI "double " "plane" ,
.BI "double " "line" ,
.BI "double " "kol" );
.SH ARGUMENTS
.TS
tab(^);
lI l.
obj^Woolz domain object with grey values.
dstErrNum^Destination pointer for error number.
gVWSp^Grey value work space.
plane^Plane (z) coordinate.
line^Line (y) coordinate.
kol^Column (x) coordinate.
gPtr^Destination pointer for a valid pointer to the grey value
^at the given coordinate.
gVal^Destination pointer for the connected grey values about
^the given coordinate.
.TE
.SH MT-LEVEL
.LP
Safe
.SH DESCRIPTION
.LP
.B WlzGreyValueMakeWSp
constructs a work space data structure which can be used for
random access to grey values in a Woolz 2D or 3D domain object.
No fields in a work space data structure should be written to
outside of the WlzGreyValue functions (except for gPtr[0-7] and gVal[0-7]).
The data structure is:
.in +4m
.ps -2
.cs B 24
.cs I 24
.cs R 24
.nf

typedef struct _WlzGreyValueWSpace
{
  WlzObjectType objType;        /* Type of object: Always either
                                   WLZ_2D_DOMAINOBJ or
                                   WLZ_3D_DOMAINOBJ */
  WlzDomain     domain;         /* The object's domain */
  WlzValues     values;         /* The object's values */
  WlzAffineTransform *invTrans; /* If the object is a WLZ_TRANS_OBJ
                                   then the inverse transform is
                                   non\-NULL */
  WlzIntervalDomain *iDom2D;    /* Current/last plane or 2D object
                                   domain */
  WlzValues     values2D;       /* Current/last plane or 2D object
                                   values */
  int           plane;          /* Current/last plane position */
  WlzGreyType   gType;          /* Grey type */
  WlzObjectType gTabType2D;     /* Current/last plane or 2D grey
                                   table */
  WlzGreyV      gBkd;           /* Background grey value, always
                                   of gType */
  WlzGreyP      gPtr[8];        /* One, four or eight grey
                                   pointers */
  WlzGreyV      gVal[8];        /* One, four or eight grey
                                   values */
} WlzGreyValueWSpace;

.fi
.cs B
.cs I
.cs R
.ps +2
.in -4m
.LP
.B WlzGreyValueFreeWSp
frees the work space data structure created by
\fBWlzGreyValueMakeWSp\fR.
.LP
.B WlzGreyValueGet
gets a single grey value and grey pointer for the given
coordinate.
.LP
.B WlzGreyValueGetCon
gets four (2D) or eight (3D) grey values and grey pointers
for the given coordinate (\fBp\fRlane, \fBl\fRine, \fBk\fRol):
.in +4m
.TS
tab(^);
l l l l l.
gPtr[\fB0\fR]^gVal[\fB0\fR]^(\fBp\fR,^\fBk\fR,^\fBl\fR)
gPtr[\fB1\fR]^gVal[\fB1\fR]^(\fBp\fR,^\fBk + 1\fR,^\fBl\fR)
gPtr[\fB2\fR]^gVal[\fB2\fR]^(\fBp\fR,^\fBk\fR,^\fBl + 1\fR)
gPtr[\fB3\fR]^gVal[\fB3\fR]^(\fBp\fR,^\fBk + 1\fR,^\fBl + 1\fR)
gPtr[\fB4\fR]^gVal[\fB4\fR]^(\fBp + 1\fR,^\fBk\fR,^\fBl\fR)
gPtr[\fB5\fR]^gVal[\fB5\fR]^(\fBp + 1\fR,^\fBk + 1\fR,^\fBl\fR)
gPtr[\fB6\fR]^gVal[\fB6\fR]^(\fBp + 1\fR,^\fBk\fR,^\fBl + 1\fR)
gPtr[\fB7\fR]^gVal[\fB7\fR]^(\fBp + 1\fR,^\fBk + 1\fR,^\fBl + 1\fR)
.TE
.in -4m
.LP
In all the above random access functions,
if the given coordinate lies outside
of the objects domain,
then the grey value is set to the background value
and the grey pointer is set to point the grey value in
the work space. If the coordinate (\fBp\fR, \fBl\fR, \fBk\fR)
lies outside the object's domain:
.in +4m
.ps -2
.cs R 24
.nf
gPtr[0] == &(gVWSp->gVal[0])
.fi
.cs R
.ps +2
.in -4m

.SH EXAMPLES
.LP
.ps -2
.cs B 24
.cs R 24
.nf

/* Simple example which reads a Woolz 2D domain object with grey
 * values from the standard input and writes a pgm format image
 * to the standard output.
 * \s+2\fBThis is NOT an efficient way to do this!\s-2\fR
 * Look at WlzCutObjToBox for a more efficient solution.
 */
 
#include <stdio.h>
#include <stdlib.h>
#include <Wlz.h>
 
int             main(char *argv[], int argc)
{
  int           line,
                kol;
  WlzObject     *obj = NULL;
  WlzGreyValueWSpace *gVWSp = NULL;
  WlzPixelV     pix;
  WlzErrorNum   errNum = WLZ_ERR_NONE;
 
  if((obj = WlzAssignObject(WlzReadObj(stdin, &errNum), NULL)) == NULL)
  {
    (void )fprintf(stderr,
                   "%s: failed to read object.\\n",
                   argv[0]);
  }
  else if((obj->type != WLZ_2D_DOMAINOBJ) ||
          (obj->domain.core == NULL) ||
          (obj->values.core == NULL))
  {
    (void )fprintf(stderr,
                   "%s: Inappropriate object type.\\n",
                   argv[0]);
    errNum = INVALID_OBJECT_TYPE;
  }
  else if((gVWSp = WlzGreyValueMakeWSp(obj, &errNum)) == NULL)
  {
    (void )fprintf(stderr,
                   "%s: Failed to make grey value work space.\\n",
                   argv[0]);
  }
  else
  {
    pix.type = gVWSp->gType;
    (void )printf("P2\\n%d %d\\n255\\n",
                  obj->domain.i->lastkl - obj->domain.i->kol1 + 1,
                  obj->domain.i->lastln - obj->domain.i->line1 + 1);
    for(line = obj->domain.i->lastln;
        line >= obj->domain.i->line1; --line)
    {
      for(kol = obj->domain.i->kol1; 
          kol <= obj->domain.i->lastkl; ++kol)
      {
        WlzGreyValueGet(gVWSp, 0, line, kol);
        pix.v = gVWSp->gVal[0];
        (void )WlzValueConvertPixel(&pix, pix, WLZ_GREY_UBYTE);
        (void )printf("%d ", pix.v.ubv);
      }
      (void )printf("\\n");
    }
    WlzGreyValueFreeWSp(gVWSp);
  }
  if(obj)
  {
    WlzFreeObj(obj);
  }
  return(errNum);
}

.fi
.cs R
.cs B
.ps +2
.SH SEE ALSO
WlzCutObjToBox(3),
WlzError(3)
.SH BUGS
Still to be found.


ChangeLog
$Log$
Revision 1.2  2007/12/05 13:33:30  bill
*** empty log message ***



binWlz/WlzGreyValue.c
#if defined(__GNUC__)
#ident "University of Edinburgh $Id$"
#else
static char _WlzGreyValue_c[] = "University of Edinburgh $Id$";
#endif
/*!
* \file         binWlz/WlzGreyValue.c
* \author       Bill Hill
* \date         May 2004
* \version      $Id$
* \par
* Address:
*               MRC Human Genetics Unit,
*               MRC Institute of Genetics and Molecular Medicine,
*               University of Edinburgh,
*               Western General Hospital,
*               Edinburgh, EH4 2XU, UK.
* \par
* Copyright (C), [2012],
* The University Court of the University of Edinburgh,
* Old College, Edinburgh, UK.
* 
* This program is free software; you can redistribute it and/or
* modify it under the terms of the GNU General Public License
* as published by the Free Software Foundation; either version 2
* of the License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be
* useful but WITHOUT ANY WARRANTY; without even the implied
* warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
* PURPOSE.  See the GNU General Public License for more
* details.
*
* You should have received a copy of the GNU General Public
* License along with this program; if not, write to the Free
* Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
* Boston, MA  02110-1301, USA.
* \brief	Gets a single grey value in an object.
* \ingroup	BinWlz
*
* \par Binary
* \ref wlzgreyvalue "WlzGreyValue"
*/

/*!
\ingroup BinWlz
\defgroup wlzgreyvalue WlzGreyValue
\par Name
WlzGreyValue - gets a single grey value in an object.
\par Synopsis
\verbatim
WlzGreyValue [-v#] [-x#] [-y#] [-z#] [-h] [<in object>]
\endverbatim
\par Options
<table width="500" border="0">
  <tr> 
    <td><b>-h</b></td>
    <td>Help, prints usage message.</td>
  </tr>
  <tr> 
    <td><b>-v</b></td>
    <td>Grey value to be set, default false.</td>
  </tr>
  <tr> 
    <td><b>-</b></td>
    <td>Column position, default 0.</td>
  </tr>
  <tr> 
    <td><b>-y</b></td>
    <td>Line position.</td>
  </tr>
  <tr> 
    <td><b>-z</b></td>
    <td>Plane position.</td>
  </tr>
  <tr> 
    <td><b>-o</b></td>
    <td>Output file name.</td>
  </tr>
</table>
\par Description
Either sets or gets a grey value within the given objects values.
If the -v flag is given the objects value is set and the modified
object is written to the output file, but if the flag is not set
then the grey value is written out in ascii form with a new line
character.
\par Examples
\verbatim
WlzGreyValue -x 100 -y 100 toucan.wlz
Outputs as ascii the value of the pixel at coordinate (100,100).
\endverbatim
\par File
\ref WlzGreyValue.c "WlzGreyValue.c"
\par See Also
\ref BinWlz "WlzIntro(1)"
\ref WlzGreyValueGet "WlzGreyValueGet(3)"
*/

#ifndef DOXYGEN_SHOULD_SKIP_THIS
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <Wlz.h>

extern int      getopt(int argc, char * const *argv, const char *optstring);
 
extern char     *optarg;
extern int      optind,
                opterr,
                optopt;

int             main(int argc, char **argv)
{
  int		option,
		setVal = 0,
  		ok = 1,
		usage = 0;
  WlzDVertex3	pos;
  WlzPixelV	pix0,
  		pix1;
  WlzGreyValueWSpace *gVWSp = NULL;
  char		*outFileStr,
  		*inObjFileStr;
  WlzErrorNum	errNum = WLZ_ERR_NONE;
  FILE		*fP = NULL;
  WlzObject	*inObj = NULL;
  const char	*errMsg;
  static char	optList[] = "o:v:x:y:z:h",
		outFileStrDef[] = "-",
  		inObjFileStrDef[] = "-";

  pos.vtX = 0;
  pos.vtY = 0;
  pos.vtZ = 0;
  opterr = 0;
  pix0.type = WLZ_GREY_DOUBLE;
  pix0.v.dbv = 0.0;
  inObjFileStr = inObjFileStrDef;
  outFileStr = outFileStrDef;
  while(ok && (usage == 0) &&
  	((option = getopt(argc, argv, optList)) != -1))
  {
    switch(option)
    {
      case 'o':
        outFileStr = optarg;
	break;
      case 'v':
	setVal = 1;
        if(sscanf(optarg, "%lg", &(pix0.v.dbv)) != 1)
	{
	  usage = 1;
	  ok = 0;
	}
	break;
      case 'x':
        if(sscanf(optarg, "%lg", &(pos.vtX)) != 1)
	{
	  usage = 1;
	  ok = 0;
	}
	break;
      case 'y':
        if(sscanf(optarg, "%lg", &(pos.vtY)) != 1)
	{
	  usage = 1;
	  ok = 0;
	}
	break;
      case 'z':
        if(sscanf(optarg, "%lg", &(pos.vtZ)) != 1)
	{
	  usage = 1;
	  ok = 0;
	}
	break;
      case 'h':
      default:
	usage = 1;
	ok = 0;
	break;
    }
  }
  if((inObjFileStr == NULL) || (*inObjFileStr == '\0') ||
     (outFileStr == NULL) || (*outFileStr == '\0'))
  {
    ok = 0;
    usage = 1;
  }
  if(ok && (optind < argc))
  {
    if((optind + 1) != argc)
    {
      usage = 1;
      ok = 0;
    }
    else
    {
      inObjFileStr = *(argv + optind);
    }
  }
  if(ok)
  {
    errNum = WLZ_ERR_READ_EOF;
    if((inObjFileStr == NULL) ||
       (*inObjFileStr == '\0') ||
       ((fP = (strcmp(inObjFileStr, "-")?
              fopen(inObjFileStr, "r"): stdin)) == NULL) ||
       ((inObj= WlzAssignObject(WlzReadObj(fP, &errNum), NULL)) == NULL))
    {
      ok = 0;
      (void )WlzStringFromErrorNum(errNum, &errMsg);
      (void )fprintf(stderr,
                     "%s: Failed to read object from file %s (%s).\n",
                     *argv, inObjFileStr, errMsg);
    }
    if(fP && strcmp(inObjFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(ok)
  {
    gVWSp = WlzGreyValueMakeWSp(inObj, &errNum);
    if(errNum != WLZ_ERR_NONE)
    {
      ok = 0;
      (void )WlzStringFromErrorNum(errNum, &errMsg);
      (void )fprintf(stderr,
		     "%s: Failed to make workspace (%s).\n",
		     *argv, errMsg);
    }
  }
  if(ok)
  {
    WlzGreyValueGet(gVWSp, pos.vtZ, pos.vtY, pos.vtX);
    if(setVal)
    {
      errNum = WlzValueConvertPixel(&pix1, pix0, gVWSp->gType);
      if(errNum == WLZ_ERR_NONE)
      {
	switch(gVWSp->gType)
	{
	  case WLZ_GREY_LONG:
	    *(gVWSp->gPtr[0].lnp) = pix1.v.lnv;
	    break;
	  case WLZ_GREY_INT:
	    *(gVWSp->gPtr[0].inp) = pix1.v.inv;
	    break;
	  case WLZ_GREY_SHORT:
	    *(gVWSp->gPtr[0].shp) = pix1.v.shv;
	    break;
	  case WLZ_GREY_UBYTE:
	    *(gVWSp->gPtr[0].ubp) = pix1.v.ubv;
	    break;
	  case WLZ_GREY_FLOAT:
	    *(gVWSp->gPtr[0].flp) = pix1.v.flv;
	    break;
	  case WLZ_GREY_DOUBLE:
	    *(gVWSp->gPtr[0].dbp) = pix1.v.dbv;
	    break;
	  case WLZ_GREY_RGBA:
	    *(gVWSp->gPtr[0].rgbp) = pix1.v.rgbv;
	    break;
	  default:
	    errNum = WLZ_ERR_GREY_TYPE;
	    break;
	}
      }
    }
    else
    {
      pix1.type = gVWSp->gType;
      pix1.v = gVWSp->gVal[0];
      errNum = WlzValueConvertPixel(&pix0, pix1, WLZ_GREY_DOUBLE);
    }
  }
  if(ok)
  {
    errNum = WLZ_ERR_WRITE_EOF;
    if(setVal == 0)
    {
      if(((fP = (strcmp(outFileStr, "-")?
		fopen(outFileStr, "w"):
		stdout)) == NULL) ||
	 (fprintf(fP, "%g\n", pix0.v.dbv) < 2))
      {
	ok = 0;
	(void )WlzStringFromErrorNum(errNum, &errMsg);
	(void )fprintf(stderr,
		       "%s: Failed to write output file (%s).\n",
		       *argv, errMsg);
      }
    }
    else
    {
      if(((fP = (strcmp(outFileStr, "-")?
		fopen(outFileStr, "w"):
		stdout)) == NULL) ||
	 ((errNum = WlzWriteObj(fP, inObj)) != WLZ_ERR_NONE))
      {
	ok = 0;
	(void )WlzStringFromErrorNum(errNum, &errMsg);
	(void )fprintf(stderr,
		       "%s: Failed to write output object (%s).\n",
		       *argv, errMsg);
      }
    }
    if(fP && strcmp(outFileStr, "-"))
    {
      fclose(fP);
    }
  }
  if(gVWSp)
  {
    WlzGreyValueFreeWSp(gVWSp);
  }
  if(inObj)
  {
    (void )WlzFreeObj(inObj);
  }
  if(usage)
  {
    (void )fprintf(stderr,
    "Usage: "
    "%s [-v#] [-x#] [-y#] [-z#] [-h] [<in object>]\n"
    "Either sets or gets a grey value within th given objects values.\n"
    "If the -v flag is given the objects value is set and the modified\n"
    "object is written to the output file, but if the flag is not set\n"
    "then the grey value is written out in ascii form with a new line\n"
    "character.\n"
    "Version: %s\n"
    "Options:\n"
    "  -v#   Grey value to be set (set to %g).\n"
    "  -x#   Column position (set to %g).\n"
    "  -y#   Line position (set to %g).\n"
    "  -z#   Plane position (set to %g).\n"
    "  -o#   Output file name.\n"
    "  -h    Display this usage information.\n",
    *argv,
    WlzVersion(),
    pix0.v.dbv,
    pos.vtX, pos.vtY, pos.vtZ);
  }
  return(!ok);
}
#endif /* DOXYGEN_SHOULD_SKIP_THIS */
